frame/chaos/src/lib.rs

```diff
@@ -0,0 +1,501 @@
+				nft.penalize_early_unstake_amount(now.as_secs(), EarlyUnstakePenalty::<T>::get())?;
+
+			// Transfer back the (possibly penalized) staked amount.
+			let chaos_account = Self::account_id();
+			// NOTE(hussein-aitlahcen): no need to keep protocol account alive.
+			T::ChaosAsset::transfer(&chaos_account, &to, penalized_stake, false)?;
+
+			// Decrement total shares.
+			// TODO: decrement by penalized only????? the protocol would own this fraction
+			// indefinitely
+			TotalShares::<T>::try_mutate(|x| {
+				x.checked_sub(nft.stake.saturated_into::<u128>().into())
+					.ok_or(ArithmeticError::Underflow)
+			})?;
+
+			// Actually burn the NFT from the storage.
+			T::burn_protocol_nft::<ChaosStakingNFTOf<T>>(instance_id)?;
+
+			// Trigger event
+			Self::deposit_event(Event::<T>::Unstaked {
+				to: to.clone(),
+				stake: nft.stake,
+				penalty: penalty_amount,
+				nft: *instance_id,
+			});
+
+			Ok(())
+		}
+
+		fn transfer_reward(
+			from: &Self::AccountId,
+			asset: Self::AssetId,
+			amount: Self::Balance,
+			keep_alive: bool,
+		) -> DispatchResult {
+			// Make sure the asset has been registered as reward, otherwise we don't track it.
+			ensure!(RewardAssets::<T>::get().contains(&asset), Error::<T>::RewardAssetDisabled);
+
+			// Transfer the reward locally.
+			let chaos_account = Self::account_id();
+			T::Assets::transfer(asset, from, &chaos_account, amount, keep_alive)?;
+
+			// Increment the reward index, used to compute user rewards.
+			RewardIndexes::<T>::try_mutate(asset, |x| {
+				let lifted_amount = RewardIndex::from(amount.saturated_into::<u128>());
+				match x {
+					Some(index) => {
+						*index =
+							(*index).checked_add(lifted_amount).ok_or(ArithmeticError::Overflow)?;
+					},
+					None => {
+						*x = Some(lifted_amount);
+					},
+				}
+				Ok(())
+			})
+		}
+
+		fn claim(to: &Self::AccountId, instance_id: &Self::InstanceId) -> DispatchResult {
+			T::try_mutate_protocol_nft(instance_id, |nft: &mut ChaosStakingNFTOf<T>| {
+				// NOTE(hussein-aitlahcen): user is still able to claim even if his position
+				// 'expired', do we allow that?
+
+				let share = U256::from(nft.stake.saturated_into::<u128>());
+				let total_shares = TotalShares::<T>::get();
+
+				// TODO(hussein-aitlahcen): extract pure maths to their own functions
+				let compute_reward = |delta_index: U256| -> Result<BalanceOf<T>, DispatchError> {
+					delta_index
+						.checked_mul(share)
+						.and_then(|x| x.checked_div(total_shares))
+						.ok_or(ArithmeticError::Overflow.into())
+						.and_then(|x| {
+							TryFrom::<U256>::try_from(x)
+								.map_err(|_| ArithmeticError::Overflow.into())
+						})
+				};
+
+				let rewards = nft
+					.reward_indexes
+					.iter()
+					.map(|(asset, index)| -> Result<(AssetIdOf<T>, BalanceOf<T>), DispatchError> {
+						match RewardIndexes::<T>::get(&asset) {
+							Some(current_index) => {
+								let delta_index = current_index.saturating_sub(*index);
```

Comment:
some mistake seems . may be should be 

>>> initial_index = 100
>>> final_index = 300
>>> share = 10
>>> total = 100
>>> (final_index-initial_index)*(share/total)
20.0

Reply:
(final_index-initial_index)*(share*(1+(duration_days/max_time)**2)/total) <- make incetive

Reply:
<img width="1145" alt="image" src="https://user-images.githubusercontent.com/757125/160500252-b1c5964e-7ba1-4b75-bb57-efeaddbf2b59.png">

in current impl any long term locker is penalized by reduction of hist share by newcomers

Reply:
anywa, current design is unsafe in terms of diluting share of long term holders too much without making short term stakers no so profitable. there is risk of long term stakers to be delluted too much.

Reply:
may be , define StakeStep (equal to 1 day). Than use formula like in WeightToFeePolynomial. So that insteaf of 1*x\**2,  can do term*x\**power for for several values, e.g. 4. Than can fit non decreasing curve to fit parameters from https://math.stackexchange.com/questions/466602/spline-interpolation-that-is-non-decreasing-when-given-non-decreasing-sequence . so we do not have to define ladder, but fit what ever is desired

Reply:
The epoch mechanism will make this issue less relevant